### PR TITLE
[WIP] Multiple Codec Support and General cleanup

### PR DESCRIPTION
**Pull request by mcdonnelldean**
Ok @mcollina,

Unfortunately this one got a bit messy because I was trying to clean up and implement multi codec support. However, I think the result is good in the end, it's broken down a bit more, names are cleaned up too. Still need some tests but provisionally, the following are done, #6 and #7 also #13 and #14 should now be easier.

### CODE REVIEW

File: hyperemitter.js [hunk 1]
```diff
@@ -58,8 +58,8 @@ function HyperEmitter (db, schema, opts) {
   })
 
   function process (change, enc, next) {
```
Comment (mcdonnelldean) [inline hyperemitter.js]: I removed the call to headers directly. We don't expose .messages (.codecs) so there was no reason to use one over the other. At this point they are merged anyway.

File: hyperemitter.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ function HyperEmitter (db, schema, opts) {
     return new HyperEmitter(db, schema, opts)
   }
 
-  this.messages = protobuf(schema)
+  this.codecs = protobuf(schema)
```
Comment (mcollina) [inline hyperemitter.js]: let's just stop accepting a schema and wrapping protobuf, we will just support codecs, and otherwise we emit buffers.

Comment (mcdonnelldean) [inline hyperemitter.js]: I think this makes the most sense. On it now

Comment (mcollina) [inline hyperemitter.js]: On second thought, I don't want to bump the major now. So, if the argument
is a string, keep the old behavior, but add a deprecation warnig. And add
an alias from 'codecs' to messages. You can use an ES5 object property for
that, and add a deprecation warning too.
Il giorno dom 24 mag 2015 alle 12:50 Dean McDonnell <
notifications@github.com> ha scritto:

Comment (mcdonnelldean) [inline hyperemitter.js]: Cool, I don't understand the first part though. If which argument is a string? I presume you mean ensure we can add the protobuf schema so we don't break existing peeps? The new sig is slightly different. I'd probably lease schema there then if you are worried about breaking things?

Comment (mcollina) [inline hyperemitter.js]: Yeah, leave schema there, but make it fully optional.
Il giorno dom 24 mag 2015 alle 13:09 Dean McDonnell <
notifications@github.com> ha scritto:

Comment (mcdonnelldean) [inline hyperemitter.js]: :thumb:

**Comment by mcdonnelldean** [comment]
@mcollina Just wanted to check if I'm on the right track here?

File: hyperemitter.js [hunk 3]
```diff
@@ -21,12 +21,20 @@ var defaults = {
...
-    return new HyperEmitter(db, schema, opts)
+    return new HyperEmitter(db, codecs, opts)
+  }
+
+  this.codecs = codecs
+  this.messages = this.codecs
+  this.status = new EventEmitter()
+
+  if (Buffer.isBuffer(this.codecs)) {
+    this.status.emit('buffers are being depreciated')
```
Comment (mcollina) [inline hyperemitter.js]: Don't emit, just use `console.log`, or better https://nodejs.org/api/util.html#util_util_deprecate_function_string.

You should also check for strings, not only Buffers.

Also, there is a spelling mistake :P.

File: hyperemitter.js [hunk 4]
```diff
@@ -106,21 +117,17 @@ function HyperEmitter (db, schema, opts) {
...
-  var err
+  var encoder = this.codecs[name]
 
-  if (!encoder) {
-    err = new Error('Non supported event')
-    if (cb) { return cb(err) }
-    else throw err
-  }
+  if (encoder) data = encoder.encode(data)
+  else this.status.emit('no encoder found for ' + name)
```
Comment (mcollina) [inline hyperemitter.js]: don't emit something like this.

I think that:
1. if the argument is a String or a Buffer, then we can send them through as Buffers, and emit them as such.
2. if it's an object, then we should return an error, just as before.

File: hyperemitter.js [hunk 5]
```diff
@@ -140,6 +147,20 @@ HyperEmitter.prototype.on = function (name, callback) {
...
+    this.codecs[name] = codec
+    return
+  }
+
+  if (Array.isArray(name)) {
+    var that = this
+    name.forEach(function (set) {
+      that.codecs[set.name] = set.codec
+    })
+  }
```
Comment (mcollina) [inline hyperemitter.js]: missing `return this` at the end. I want a chainable API.

File: hyperemitter.js [hunk 6]
```diff
@@ -58,13 +66,16 @@ function HyperEmitter (db, schema, opts) {
...
-  }
+    var container = that.codecs.Event.decode(change.value)
+    var name = container.name
+    var decoder = that.codecs[name]
+    var event = container.payload
 
-  // the status of this HyperEmitter
-  this.status = new EventEmitter()
+    if (decoder) event = decoder.decode(event)
+    else that.status.emit('no decoder found for ' + name)
```
Comment (mcollina) [inline hyperemitter.js]: Instead of emitting a status message, just emit the event name with the Buffer as an argument.

**Comment by mcollina** [comment]
You are on the right track. Be sure to update tests along the way, as I think there are some functionality changes.

File: hyperemitter.js (hunk 3 above)
Comment (mcdonnelldean) [inline hyperemitter.js]: @mcollina What should I do with a string? Is that an acceptable input to Protobuf?

File: hyperemitter.js (hunk 4 above)
Comment (mcdonnelldean) [inline hyperemitter.js]: Noted

File: hyperemitter.js (hunk 5 above)
Comment (mcdonnelldean) [inline hyperemitter.js]: noted

File: hyperemitter.js (hunk 6 above)
Comment (mcdonnelldean) [inline hyperemitter.js]: noted.

**Comment by mcdonnelldean** [comment]
@mcollina Awesome, yeah this was first pass, I have a few tests to write to complete. On potential issues or maybe just a question.

If we have an codec we will encode / decode. If we don't have it we wont. There is a potential issue in that the payload may be encoded on one side but missing the decoder on the other side. Should the event container specify if it's payload is encoded or not?

Also, should we have a flag that allows the user to turn on missing encoder / decoder messages. I feel that a lot of the status.emit messages would be better as log messages.

**Comment by mcollina** [comment]
I think we should have a `unknownEvent` handler in the options, with a default.

> If we have an codec we will encode / decode. If we don't have it we wont. There is a potential issue in that the payload may be encoded on one side but missing the decoder on the other side. Should the event container specify if it's payload is encoded or not?

I don't think so. If that happens it's either a user-wanted thing or not, and we can control it via a construction flag. There are two cases:
1. the user is running two different codebases, so they possibly have different events.
2. the user is running two versions of the same codebase, and the new one has new events.

So, in case one the user wants the 'ignore' behavior (decode as Buffers), while in the second it might want the 'error' behavior, so that the process is killed.

Either 1 or 2 should be the default. I'm more inclined to 1 (do not error, but risk losing messages).

BTW, do not put 'phrases' in the events, but use a single word.

File: hyperemitter.js (hunk 3 above)
Comment (mcollina) [inline hyperemitter.js]: Yes it is.

**Comment by mcdonnelldean** [comment]
@mcollina. Could you review whenever you get a chance. Cheers 😊

**Comment by mcollina** [comment]
This is awesome! Can you please add a couple of tests for the registerCodec method, and document that one in the README?

**Comment by mcdonnelldean** [comment]
Will do, ill ping you later when they are done

**Comment by mcdonnelldean** [comment]
@mcollina Done :dancer:

**Comment by mcollina** [comment]
Whoaaa!!

**Comment by mcollina** [comment]
Released as 1.2.0!

**Pull request merged**